### DISCUSSION TITLE
Installation and build failure on Ubuntu

### DISCUSSION THREAD

**Discussion by OM-BDF** [rdkit/rdkit#6517] [Q&A]
Hello RDKit community,

First of all, I did search for similar discussions, without success. My apologies in advances if I did miss something.

I am trying to prepare a PostgreSQL system with the RDKit cartridge.

I do have some basic backgrounds on terminal commands but I do not know the deep inside of Linux system.

I tried to prepare this system on a macOS 12.6.7 Monterey and mainly on a fresh install of Ubuntu 22.04.
In both cases, I did follow religiously the instructions described in https://rdkit.org/docs/Install.html, but I encountered the same errors.

Each time, I used the latest available versions of software. I was for example making sure that Conda was updated and upgraded. 

-- I first tried the install it using Conda. Conda and RDKit installations went fine. 
However, when I run ‘conda install -c rdkit rdkit-postgresql‘, I got the following errors:

UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:

Specifications:

  - contourpy -> python[version='>=2.7,<2.8.0a0|>=3.6,<3.7.0a0']
  - fonttools -> python[version='>=3.10,<3.11.0a0|>=3.7,<3.8.0a0|>=3.9,<3.10.0a0|>=3.8,<3.9.0a0|>=3.6,<3.7.0a0']
  - python-dateutil -> python[version='>=3.8,<3.9.0a0|>=3.9,<3.10.0a0']
  - rdkit -> python[version='>=3.10,<3.11.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0|>=3.7.1,<3.8.0a0']
  - rdkit-postgresql -> python[version='2.7.*|3.4.*|3.5.*']
  - six -> python[version='>=2.7,<2.8.0a0|>=3.6,<3.7.0a0|>=3.8,<3.9.0a0|>=3.7,<3.8.0a0|>=3.9,<3.10.0a0|>=3.5,<3.6.0a0']

Your python: python=3.11

If python is on the left-most side of the chain, that's the version you've asked for.
When python appears to the right, that indicates that the thing on the left is somehow
not available for the python version you are constrained to. Note that conda will not
change your python version to a different minor version unless you explicitly specify
that.

The following specifications were found to be incompatible with a past
explicit spec that is not an explicit spec in this operation (rdkit):

- rdkit-postgresql -> rdkit[version='2014.03.1|2014.09.1|2014.09.2|2015.03.1|2015.09.2|>=2016.03.1|>=2016.03.2|>=2016.03.3|>=2016.03.4,<2016.09']

Follows a long list of packages names with versions numbers. 
Unfortunately, I do not know how to interpret this log. Could someone hint me what could be wrong here?

-- Secondly, I made a fresh install of the Ubuntu 22.04. I tried this time to build everything. I installed Conda, cmake, boost, gxx-linyx-64, numpy, matplotlib, git. I got the rdkit github repository, create a Conda environment, activate it. When I tried to run the cmake command, I got an error, telling me that cmake cannot compile a simple program.

Again, could someone give me hints here.

-- Finally, I tried via the Ubuntu repositories or the Homebrew formula.
I also tried the procedure described in the page https://github.com/rdkit/conda-rdkit again on a fresh install of Ubuntu 22.04.

In all cases, never the postgreSQL RDKit cartridge became available. The initdb command for example was never available in the ‘my-env/bin’ folder.

I also tried to follow all advices provided in the ‘/docs/install’ page. Without success.

I would greatly appreciate if someone could help me and give me some hints on what could be wrong.
Thank you in advance.

**Comment by ptosco** [comment]
This blog post has step-by-step instructions to build a conda-based RDKit installation from source on Linux and Mac:
https://greglandrum.github.io/rdkit-blog/posts/2023-03-17-setting-up-a-cxx-dev-env2.html

Once you have the above, you can follow the step-by-step instructions in this README to build the PostgreSQL cartridge matching your RDKit installation:
https://github.com/rdkit/rdkit/tree/master/Code/PgSQL/rdkit#readme

**Comment by OM-BDF** [comment]
Dear Paolo,

thank you very much for your links.
Following those, I was more successful building/installing the RDKit.
But, at the last step, the installation of the RDKit cartridge in the postgreSQL database, I got an error I do not how to solve:

postgres@Ubu:/etc/postgresql/15/main$ psql -c 'create extension rdkit' --port 5433 pgdbol
2023-07-05 15:54:15.788 CEST [464334] ERROR:  could not load library "/usr/lib/postgresql/15/lib/rdkit.so": libboost_serialization.so.1.82.0: cannot open shared object file: No such file or directory

2023-07-05 15:54:15.788 CEST [464334] STATEMENT:  create extension rdkit
ERROR:  could not load library "/usr/lib/postgresql/15/lib/rdkit.so": libboost_serialization.so.1.82.0: cannot open shared object file: No such file or directory

postgres@Ubu:/etc/postgresql/15/main$ echo $LD_LIBRARY_PATH
/home/olivier/Documents/RDKit_Git/lib:/home/olivier/anaconda3/envs/py310_rdkit_build/lib
postgres@Ubu:/etc/postgresql/15/main$ 


I tried with postgres account, my account, and with a sudo command.
Same error.
I tried with the server running or stopped, same error.

I checked the LIBRARY_PATH. In the anaconda3/envs/… one, the libboost_serialization.so.1.82.0 file does exist! With a chmod of 775.
I tried to change its ownership, same error.

With all that, I reached the limit of my knowledge. Could you hint me regarding this problem ?
Thank you in advance.

Best regards,
Olivier

**Comment by ptosco** [comment]
Dear Olivier,

check [this paragraph](https://github.com/rdkit/rdkit/tree/master/Code/PgSQL/rdkit#cartridge-installation-into-non-conda-postgresql) in my README about `LD_LIBRARY_PATH`.
You need to add an entry to your `/etc/postgresql/15/main/environment` (you will need `sudo` privileges):
```
$ echo "LD_LIBRARY_PATH = '/home/olivier/Documents/RDKit_Git/lib:/home/olivier/anaconda3/envs/py310_rdkit_build/lib'" | sudo tee -a /etc/postgresql/15/main/environment
```
and then restart the PostgreSQL server:
```
$ sudo systemctl restart postgresql
```
Then loading the cartridge should succeed.

**Reply by OM-BDF** [reply]
Dear Paolo,

thank you for your advise.
Unfortunately, I already did that.
Out of conscience, I did again. I open the file and this line did appear twice.

And the same error still appears.

I restarted the server. I echoed the variable, and it was still empty. I therefore explicitely create this variable with the command export. Still the same error…

See my log below:

```
(base) olivier@Ubu:~$ echo "LD_LIBRARY_PATH = '/home/olivier/Documents/RDKit_Git/lib:/home/olivier/anaconda3/envs/py310_rdkit_build/lib'" | sudo tee -a /etc/postgresql/15/main/environment
[sudo] password for olivier: 
LD_LIBRARY_PATH = '/home/olivier/Documents/RDKit_Git/lib:/home/olivier/anaconda3/envs/py310_rdkit_build/lib'
(base) olivier@Ubu:~$ conda activate py310_rdkit_build
(py310_rdkit_build) olivier@Ubu:~$ su postgres
Password: 

postgres@Ubu:~$ pg_ctl –port 5433 -D $PGDATA start 
waiting for server to start....2023-07-06 14:11:59.197 CEST [4438] LOG:  starting PostgreSQL 15.3 (Ubuntu 15.3-1.pgdg22.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0, 64-bit
2023-07-06 14:11:59.197 CEST [4438] LOG:  listening on IPv4 address "127.0.0.1", port 5433
2023-07-06 14:11:59.222 CEST [4438] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5433"
2023-07-06 14:11:59.306 CEST [4441] LOG:  database system was shut down at 2023-07-05 16:15:11 CEST
2023-07-06 14:11:59.407 CEST [4438] LOG:  database system is ready to accept connections
 done
server started

postgres@Ubu:~$ psql --port 5433 -c 'create extension rdkit' pgdbol
2023-07-06 14:13:47.560 CEST [5165] ERROR:  could not load library "/usr/lib/postgresql/15/lib/rdkit.so": libboost_serialization.so.1.82.0: cannot open shared object file: No such file or directory
2023-07-06 14:13:47.560 CEST [5165] STATEMENT:  create extension rdkit
ERROR:  could not load library "/usr/lib/postgresql/15/lib/rdkit.so": libboost_serialization.so.1.82.0: cannot open shared object file: No such file or directory
postgres@Ubu:~$ echo $LD_LIBRARY_PATH

postgres@Ubu:~$ export LD_LIBRARY_PATH="/home/olivier/Documents/RDKit_Git/lib:/home/olivier/anaconda3/envs/py310_rdkit_build/lib"
postgres@Ubu:~$ echo $LD_LIBRARY_PATH
/home/olivier/Documents/RDKit_Git/lib:/home/olivier/anaconda3/envs/py310_rdkit_build/lib
postgres@Ubu:~$ psql --port 5433 -c 'create extension rdkit' pgdbol
2023-07-06 14:16:07.559 CEST [5286] ERROR:  could not load library "/usr/lib/postgresql/15/lib/rdkit.so": libboost_serialization.so.1.82.0: cannot open shared object file: No such file or directory
2023-07-06 14:16:07.559 CEST [5286] STATEMENT:  create extension rdkit
ERROR:  could not load library "/usr/lib/postgresql/15/lib/rdkit.so": libboost_serialization.so.1.82.0: cannot open shared object file: No such file or directory
postgres@Ubu:~$ exit
```


I also explicitly export all the variables before testing again, no success neither.
```
export PGDATA="/var/lib/postgresql/data"
export PATH="/usr/lib/postgresql/15/bin:$PATH:"
export RDBASE="/home/olivier/Documents/RDKit_Git"
export PYTHONPATH="$RDBASE"
export LD_LIBRARY_PATH="$RDBASE/lib:$CONDA_PREFIX/lib"
```

I tried to reboot the machine. And still the same error.

Thank you in advance for your help.

Best regards,
Olivier

**Reply by ptosco** [reply]
Please post the output of
```
patchelf --print-rpath /usr/lib/postgresql/15/lib/rdkit.so
```
and
```
LD_LIBRARY_PATH='/home/olivier/Documents/RDKit_Git/lib:/home/olivier/anaconda3/envs/py310_rdkit_build/lib' ldd /usr/lib/postgresql/15/lib/rdkit.so
```

**Reply by OM-BDF** [reply]
Dear Paolo,

yes the `libboost_serialization.so.1.82.0` file is in `/home/olivier/anaconda3/envs/py310_rdkit_build/lib`.

Below the log of your command requests:

```
(base) olivier@Ubu:~/anaconda3/envs/py310_rdkit_build/lib$ patchelf --print-rpath /usr/lib/postgresql/15/lib/rdkit.so
/home/olivier/Documents/RDKit_git/lib:/home/olivier/anaconda3/envs/py310_rdkit_build/lib


(base) olivier@Ubu:~/anaconda3/envs/py310_rdkit_build/lib$ LD_LIBRARY_PATH='/home/olivier/Documents/RDKit_Git/lib:/home/olivier/anaconda3/envs/py310_rdkit_build/lib' ldd /usr/lib/postgresql/15/lib/rdkit.so
	linux-vdso.so.1 (0x00007ffe2b8b5000)
	libfreetype.so.6 => /home/olivier/anaconda3/envs/py310_rdkit_build/lib/libfreetype.so.6 (0x00007f129fda4000)
	libboost_serialization.so.1.82.0 => /home/olivier/anaconda3/envs/py310_rdkit_build/lib/libboost_serialization.so.1.82.0 (0x00007f129fd5f000)
	libstdc++.so.6 => /home/olivier/anaconda3/envs/py310_rdkit_build/lib/libstdc++.so.6 (0x00007f129f21d000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f129f136000)
	libgcc_s.so.1 => /home/olivier/anaconda3/envs/py310_rdkit_build/lib/libgcc_s.so.1 (0x00007f129fd33000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f129ee00000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f129fe5e000)
	libz.so.1 => /home/olivier/anaconda3/envs/py310_rdkit_build/lib/./libz.so.1 (0x00007f129fd16000)
	libpng16.so.16 => /home/olivier/anaconda3/envs/py310_rdkit_build/lib/./libpng16.so.16 (0x00007f129fcda000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f129fcd5000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f129fcd0000)
(base) olivier@Ubu:~/anaconda3/envs/py310_rdkit_build/lib$ 
```


Thank you again.
Best regards,
Olivier

**Reply by OM-BDF** [reply]
Dear Paolo,

I would like to thank you for your patience and your advices.
Following the following hints:

_sudo mkdir -p /opt/postgres/rdkit/lib
sudo cp /home/olivier/anaconda3/envs/py310_rdkit_build/lib/libboost_serialization.so.1.82.0 /opt/postgres/rdkit/lib
sudo chmod -R 755 /opt/postgres/rdkit/lib
sudo patchelf --set-rpath /opt/postgres/rdkit/lib /usr/lib/postgresql/15/lib/rdkit.so
Then the cartridge should load. You can also remove the LD_LIBRARY_PATH entries from /etc/postgresql/15/main/environment, I don't think it should be necessary._

I managed to install the RDKit extension inside my PostgreSQL 15 system.
Thank you very much.

I also managed to reproduce the setting in a MacOS system as well. 

Now, I will load molecules and play with it !

Best regards,
Olivier

**Comment by ptosco** [comment]
Could you please report the output of:
```
sudo su postgres -s /bin/bash -c 'LD_LIBRARY_PATH=/home/olivier/Documents/RDKit_Git/lib:/home/olivier/anaconda3/envs/py310_rdkit_build/lib ldd /usr/lib/postgresql/15/lib/rdkit.so'
```

**Reply by OM-BDF** [reply]
Dear Paolo,

the output is as follows:

```
_(base) olivier@Ubu:~$ sudo su postgres -s /bin/bash -c 'LD_LIBRARY_PATH=/home/olivier/Documents/RDKit_Git/lib:/home/olivier/anaconda3/envs/py310_rdkit_build/lib ldd /usr/lib/postgresql/15/lib/rdkit.so'
[sudo] password for olivier: 
    linux-vdso.so.1 (0x00007ffe01f0d000)
    libfreetype.so.6 => /lib/x86_64-linux-gnu/libfreetype.so.6 (0x00007f2c1a938000)
    libboost_serialization.so.1.82.0 => not found
    libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f2c1a600000)
    libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f2c1a851000)
    libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f2c1b2c3000)
    libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f2c1a200000)
    /lib64/ld-linux-x86-64.so.2 (0x00007f2c1b2f6000)
    libpng16.so.16 => /lib/x86_64-linux-gnu/libpng16.so.16 (0x00007f2c1a5c5000)
    libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f2c1b2a5000)
    libbrotlidec.so.1 => /lib/x86_64-linux-gnu/libbrotlidec.so.1 (0x00007f2c1b297000)
    libbrotlicommon.so.1 => /lib/x86_64-linux-gnu/libbrotlicommon.so.1 (0x00007f2c1a82e000)_
```

Indeed, the libboost_serialization file is not found.
I made the following:

```
_(py310_rdkit_build) olivier@Ubu:~$ cd anaconda3/envs/py310_rdkit_build/lib
(py310_rdkit_build) olivier@Ubu:~/anaconda3/envs/py310_rdkit_build/lib$ ls libboost_serial*
libboost_serialization.so  libboost_serialization.so.1.82.0
(py310_rdkit_build) olivier@Ubu:~/anaconda3/envs/py310_rdkit_build/lib$ ls -la libboost_serial*
lrwxrwxrwx 1 olivier olivier     32 juil.  4 19:14 libboost_serialization.so -> libboost_serialization.so.1.82.0
-rwxrwxr-x 2 olivier olivier 483920 avril 19 12:24 libboost_serialization.so.1.82.0
(py310_rdkit_build) olivier@Ubu:~/anaconda3/envs/py310_rdkit_build/lib$_
```

The file does exist in the folder. Could it be a file ownership problem ?

Thank you for your help.
Best regards,
Olivier

**Reply by ptosco** [reply]
Yes, that confirms what I suspected when I asked you to rerun `ldd` as the `postgres` user.
You may simply do something like
```
sudo mkdir -p /opt/postgres/rdkit/lib
sudo cp /home/olivier/anaconda3/envs/py310_rdkit_build/lib/libboost_serialization.so.1.82.0 /opt/postgres/rdkit/lib
sudo chmod -R 755 /opt/postgres/rdkit/lib
sudo patchelf --set-rpath /opt/postgres/rdkit/lib /usr/lib/postgresql/15/lib/rdkit.so
```
Then the cartridge should load. You can also remove the `LD_LIBRARY_PATH` entries from `/etc/postgresql/15/main/environment`, I think those shouldn't be necessary.
